File: problem_12/main.c

//
// Project Euler Problem 12: Highly divisible triangular number
//
// D. C. Groothuizen Dijkema - February, 2020
//

#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <time.h>

int triangle_number(const int n)
{
  return n*(n+1)/2;
}

int num_divisors(int n, const int init)
{
  if (n==1|n==2) { return n; }

  for (int itr=init;itr<(int)sqrt(n)+1;++itr)
  {
    // we've found the first prime that divides into n
    if (n%itr==0)
    {
      // divide the prime out of n
      int cnt=1;
      while (n%itr==0)
      {
        ++cnt;
        n/=itr;
      }
      // find the next prime that divides in, starting from after the current prime
      return cnt*num_divisors(n,itr+1);
    }
  }
  // if nothing up to sqrt(n) divides in, it must be prime
  return 2;
}

void run(void)
{
  bool found=false;
  int cnt_divisors=2;
  for (int itr=3;!found;++itr)
  {
    // itr and (itr+1) will never share factors as they are coprime.
    if (itr%2==0) { cnt_divisors=num_divisors(itr/2,2)*num_divisors(itr+1,2); }
    else { cnt_divisors=num_divisors(itr,2)*num_divisors((itr+1)/2,2); }

    if (cnt_divisors>500)
    {
      printf("The first triangle number with more than 500 factors is: %d\n",triangle_number(itr));
      found=true;
    }
  }
}

int main(void)
{
  clock_t t=clock(); 
  run();
  printf("Elapsed time: %fs\n",((double)(clock()-t))/CLOCKS_PER_SEC); 

  return 0; 
}

File: problem_12/main.ml

(* 
Project Euler Problem 12: Highly divisible triangular number

D. C. Groothuizen Dijkema - February, 2020
*)

let triangle n=
  n*(n+1)/2

let num_divisors n=
  let upper=int_of_float (floor(sqrt(float_of_int n))) in
    let rec count_divisors m init=
      if m==1||m==2 then m
      else let rec loop itr=
        if itr>upper then 2
        else if m mod itr==0 then
          let rec divide o cnt=
            if o mod itr==0 then divide (o/itr) (succ cnt)
            else (o,cnt)
          in 
          let (out,cnt)=divide m 1 in
            (count_divisors out (itr+1))*cnt
        else loop (itr+1)
      in loop init
    in count_divisors n 2

let run=
  let rec loop itr=
    let t=triangle itr in
      if itr mod 2==0 then 
        if (num_divisors (itr/2))*(num_divisors (succ itr)) > 500 then t
        else loop (succ itr)
      else
        if (num_divisors itr)*(num_divisors ((succ itr)/2)) > 500 then t
        else loop (succ itr)
  in loop 3

let a=run
let ()=print_endline (string_of_int a)
